types/moko/LW001-BG-PRO/uplink.js

```diff
@@ -0,0 +1,296 @@
+
+  } else if (port === 2) {
+    let parseLen = 3; // common head is 3 byte
+    const datas = [];
+    const positionTypeCode = bytes[parseLen++];
+
+    const year = bytes[parseLen] * 256 + bytes[parseLen + 1];
+    parseLen += 2;
+    const mon = bytes[parseLen++];
+    const days = bytes[parseLen++];
+    const hour = bytes[parseLen++];
+    const minute = bytes[parseLen++];
+    const sec = bytes[parseLen++];
+    const timezone = bytes[parseLen++];
+    let timestamp = `${year}-${mon}-${days} ${hour}:${minute}:${sec} TZ:${timezone}`;
+
+    if (timezone > 0x80) {
+      timestamp = `${year}-${mon}-${days} ${hour}:${minute}:${sec} TZ:${timezone - 0x100}`;
+    }
+
+    const datalen = bytes[parseLen++];
+    if (positionTypeCode === 0 || positionTypeCode === 1) {
+      for (let i = 0; i < (datalen / 7); i++) {
+        const tempData = {};
+        tempData.mac = substringBytes(bytes, parseLen, 6);
+        parseLen += 6;
+        tempData.rssi = bytes[parseLen++] - 256;
+        datas.push(tempData);
+      }
+      if (positionTypeCode === 0) {
+        wifi.macData = datas;
+        wifi.timestamp = timestamp;
+      } else {
+        bluetooth.macData = datas;
+        bluetooth.timestamp = timestamp;
+      }
+    } else {
+      let lat = Bits.bitsToUnsigned(bits.substr(parseLen * 8, 32));
+      parseLen += 4;
+      let lon = Bits.bitsToUnsigned(bits.substr(parseLen * 8, 32));
+      parseLen += 4;
+
+      if (lat > 0x80000000) {
+        lat -= 0x100000000;
+      }
+
+      if (lon > 0x80000000) {
+        lon -= 0x100000000;
+      }
+
+      gps.latitude = lat / 10000000;
+      gps.longitude = lon / 10000000;
+      gps.pdop = bytes[parseLen] / 10;
+      gps.timestamp = timestamp;
+    }
+  } else if (port === 3) {
+    let parseLen = 3;
+    const datas = [];
+    const failedTypeCode = Bits.bitsToUnsigned(bits.substr(parseLen * 8, 8));
+    failure.reasonsForPositioningFailure = posFailedReasonArray[failedTypeCode];
+    const datalen = bytes[parseLen++];
+    // wifi and ble reason
+    if (failedTypeCode <= 5) {
+      if (datalen) {
+        for (let i = 1; i < (datalen / 7); i++) {
+          const item = {};
+          item.mac = substringBytes(bytes, parseLen, 6);
+          parseLen += 6;
+          item.rssi = bytes[parseLen++] - 256;
+          datas.push(item);
+        }
+        failure.macData = datas;
+      }
+      // gps reason
+    } else if (failedTypeCode <= 11) {
+      const pdop = bytes[parseLen++];
+      if (pdop !== 0xff) {
+        failure.pdop = pdop / 10
+      } else {
+        failure.pdop = "UNKNOWN";
+      }
+      failure.gpsSatelliteCn = `${bytes[parseLen]}-${bytes[parseLen + 1]}-${bytes[parseLen + 2]}-${bytes[parseLen + 3]}`;
+    }
+  } else if (port === 4) {
+    shutdown.shutdownType = shutdownTypeArray[Bits.bitsToUnsigned(bits.substr(24, 8))];
+  } else if (port === 5) {
+    vibration.numberOfShocks = Bits.bitsToUnsigned(bits.substr(24, 16));
```

Comment:
possibly directly move the emits up here to the port check to prevent the additional ifs